src/core.rs

```diff
@@ -103,54 +97,86 @@ pub async fn main(cli: &Cli) -> Result<(), Error> {
     let handler_token = token.clone();
     ctrlc::set_handler(move || handler_token.cancel())?;
 
-    // Prepare TLS related stuff
-    let (rustls_cfg, custom_domain_providers) = tls::setup(
-        cli,
-        &mut tasks,
-        domains.clone(),
-        http_client.clone(),
-        Arc::new(dns_resolver),
-        tls_session_cache.clone(),
-        &registry,
-    )
-    .await
-    .context("unable to setup TLS")?;
-
-    // Create routers
-    let https_router = routing::setup_router(
-        cli,
-        custom_domain_providers,
-        &mut tasks,
-        http_client.clone(),
-        reqwest_client,
-        &registry,
-        clickhouse.clone(),
-        vector.clone(),
-    )
-    .await?;
-    let http_router = Router::new().fallback(routing::redirect_to_https);
+    // TLS session cache
+    let tls_session_cache = Arc::new(sessions::Storage::new(
+        cli.http_server.http_server_tls_session_cache_size,
+        cli.http_server.http_server_tls_session_cache_tti,
+    ));
 
     // HTTP server metrics
     let http_metrics = http::server::Metrics::new(&registry);
 
-    // Set up HTTP
-    let http_server = Arc::new(http::Server::new(
-        http::server::Addr::Tcp(cli.http_server.http_server_listen_plain),
-        http_router,
-        (&cli.http_server).into(),
-        http_metrics.clone(),
-        None,
-    ));
-    tasks.add("http_server", http_server);
-
-    let https_server = Arc::new(http::Server::new(
-        http::server::Addr::Tcp(cli.http_server.http_server_listen_tls),
-        https_router,
-        (&cli.http_server).into(),
-        http_metrics.clone(),
-        Some(rustls_cfg),
-    ));
-    tasks.add("https_server", https_server);
+    if cli.misc.insecure_serve_http_only {
```

Comment:
Maybe we can make this more concise? Like make `custom_domain_providers` parametric depending on `cli.misc.insecure_serve_http_only`.

Add some `http_router` which would also depend on the `cli.misc.insecure_serve_http_only` and be either redirect or real one.

And then run TLS related stuff in some
```
if !cli.misc.insecure_serve_http_only {

}
```
block? Just idea, but should work I guess. This would make code less redundant.

Reply:
I tried to reshuffle it a bit. Let me know what you think!